### ISSUE TITLE
Unstable methods can interfere with stable method calls which use auto-dereferencing

### ISSUE THREAD

**Issue by jturner314** [rust-lang/rust#86682]
Given the following code:

```rust
use std::ops::Range;

fn main() {
    let mut x: Range<usize> = 3..5;
    let y: &mut Range<usize> = &mut x;
    println!("{}", y.is_empty());
}
```

([Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=34698fdfbff0ffe1f6ada0bd84a2d5fe))

The current output is:

```
error[E0658]: use of unstable library feature 'exact_size_is_empty'
 --> src/main.rs:6:22
  |
6 |     println!("{}", y.is_empty());
  |                      ^^^^^^^^
  |
  = note: see issue #35428 <https://github.com/rust-lang/rust/issues/35428> for more information
```

Ideally, the example will compile without errors, since auto-deref should result in a call to the stable [`Range::is_empty`](https://doc.rust-lang.org/std/ops/struct.Range.html#method.is_empty) method. Currently, the user has to work around the error by changing `y.is_empty()` to `(*y).is_empty()`.

I suspect that the error is due to the following:

- `&mut Range<usize>` implements `ExactSizeIterator` (via [`impl<'_, I> ExactSizeIterator for &'_ mut I`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#impl-ExactSizeIterator-23) and [`impl ExactSizeIterator for Range<usize>`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#impl-ExactSizeIterator-15))
- `ExactSizeIterator` has an unstable [`is_empty`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#method.is_empty) method.
- So, the compiler sees that there is an *unstable* method named `is_empty` for `&mut Range<usize>` (via `ExactSizeIterator`), and as a result, it stops before finding the *stable* `Range::is_empty` method via auto-dereferencing.

I'd suggest changing the compiler to continue checking for applicable methods (including via auto-dereferencing) even after finding an applicable *unstable* method, so that adding unstable methods to traits doesn't interfere with existing code. (I don't think it's a regression in this particular case, since I think `ExactSizeIterator::is_empty` was added before `Range::is_empty` was stabilized, but I could easily see this compiler behavior causing regressions for other additions of unstable methods to traits in the future.)

This error occurs on stable (1.53.0), beta (1.54.0-beta.1), and nightly (1.55.0-nightly), tested via the Rust Playground.

**Comment by inquisitivecrystal** [comment]
It doesn't feel like there's a great solution here. The current state of affairs breaks the invariant that adding an unstable method shouldn't affect stable code, whereas the proposed change would break the invariant that stabilizing a method shouldn't change anything for people who have already opted into it as unstable. The latter is probably more acceptable, but neither of them seems great to me.

**Comment by jturner314** [comment]
> stabilizing a method shouldn't change anything for people who have already opted into it as unstable

Good point. There are two cases of interest:

1. The user has not opted-in to the unstable feature (i.e. they're on stable, beta, or nightly without enabling the feature).

   IMO, in this case, the compiler should ignore the unstable feature for the purpose of compiling the code. In other words, the only thing the compiler should do with not-enabled unstable features is to suggest that the user enable the relevant feature, if applicable.

   For the example above, the code would compile without errors, with a call to `Range::is_empty`. Optionally, the compiler could notify the user that if they enable the unstable `exact_size_is_empty` feature, then the behavior will change (i.e. the call will be to `ExactSizeIterator::is_empty` instead).

2. The user has opted-in to the unstable feature.
 
   In this case, the compiler should enable that functionality.

   For the example above, if the user added `#![feature(exact_size_is_empty)]`, then the code would compile without errors, with a call to `ExactSizeIterator::is_empty`. Optionally, the compiler could notify the user that if they disable the feature, then the behavior will change (i.e. the call will be to `Range::is_empty` instead).

**Comment by RodBurman** [comment]
As of release 1.83.0:
cargo 1.83.0 (5ffbef321 2024-10-29)
release: 1.83.0
commit-hash: 5ffbef3211a8c378857905775a15c5b32a174d3b
commit-date: 2024-10-29
host: aarch64-apple-darwin
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.7.1 (sys:0.4.74+curl-8.9.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 15.1.1 [64-bit]

```
The example complies with the following warning:
cargo build
   Compiling unstst v0.1.0 (/Users/rod/code/rust/triage/unstst)
warning: a method with this name may be added to the standard library in the future
 --> src/main.rs:6:22
  |
6 |     println!("{}", y.is_empty());
  |                      ^^^^^^^^
  |
  = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
  = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
  = help: call with fully qualified syntax `std::ops::Range::<Idx>::is_empty(...)` to keep using the current method
  = note: `#[warn(unstable_name_collisions)]` on by default

warning: `unstst` (bin "unstst") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.46s
```

Which I think explains the problem well and we can close this issue?

**Comment by jturner314** [comment]
Thanks! Yes, this issue is resolved in 1.83.
